20250115-actors-bi-di-stream.md

```diff
@@ -0,0 +1,182 @@
+This requires the app to both have an open port, and be routable by Daprd to receive these requests.
+Additionally, there is no current mechanism for dynamic Actor Type discovery.
+
+By implementing the Actor API over a gRPC bi-directional stream, the app does not need to open an HTTP port to receive Actor invocations.
+This approach also enables dynamic Actor Type registration.
+
+## Implementation Details
+
+The protos that implement the new Actor stream runtime reside in the new proto package `dapr.proto.actors.v1`.
+
+```proto
+// Actors service provides APIs for user applications to interact with the Dapr Actor
+// runtime and receive actor messages.
+service Actors {
+  // Stream is the bi-directional streaming RPC for actors to receive messages
+  // and send responses.
+  rpc Stream(stream Request) returns (stream Response) {}
+}
+```
+
+### Initial Message
+
+Each gRPC stream connected to Daprd will first register its Actor Type along with that type’s entity configuration.
+Specifically, the following fields for that type will be included:
+
+```go
+type EntityConfig struct {
+    ActorIdleTimeout           string
+    DrainOngoingCallTimeout    string
+    DrainRebalancedActors      bool
+    Reentrancy                 ReentrancyConfig
+    RemindersStoragePartitions int
+}
+```
+
+Upon successful registration, which involves Daprd registering the Actor Type with the Actor runtime, the stream will be used to send and receive Actor invocations.
+Registration involves adding that type to the internal actor table factories, targeting the gRPC app stream, and broadcasting the hosted actor type to placement.
+
+Similar to the `SubscribeTopicEventsAlpha1` RPC, the Actors runtime will take a `oneof` request message, with an initial message resembling `[SubscribeTopicEventsRequestAlpha1](https://github.com/dapr/dapr/blob/776209e56e12fbff4bac68fad0c7f28e0eaf6ec2/dapr/proto/runtime/v1/dapr.proto#L450)`.
+
+If multiple initial streams of the same Actor Type are registered, they must all contain the same initial message; otherwise, subsequent streams will be rejected.
+The first message from a client on a stream must be the initial message.
+The initial request will contain the following message:
+
+```proto
+// RequestInitial is the initial message containing details for
+// configuring actor runtime APIs.
+message RequestInitial {
+  string type = 1; // Required: Actor Type
+  optional google.protobuf.Duration idle_timeout = 2;
+  optional bool drain_rebalanced = 3;
+  optional google.protobuf.Duration drain_ongoing_call_timeout = 4;
+  optional RequestInitialReentrancyConfig reentrancy_config = 5;
+}
+
+message RequestInitialReentrancyConfig {
+  optional int32 max_stack_depth = 1;
+}
+```
+
+### Actor API Requests
+
+After the initial message, the client sends API requests to Daprd or processes received message events.
+
+```proto
+message Request {
+  oneof actor_request_type {
```

Comment:
There should be an error channel exposed here somewhere so that if the client requests the creation of an actor and some prerequisite on the runtime isn't met (e.g. cannot connect to the specified actor state store), such a message can be sent back to the SDK so an exception can be raised.

Reply:
Sending the error response on initialization should be covered in the [`ResponseInitial` message](https://github.com/dapr/proposals/pull/72/files#diff-91f410445466f13b1f41aa800ad691ac699a1a487482db8366548b14a7b72964R137) sent back to the client.

Reply:
This will be covered by the [`ResponeResponse` message](https://github.com/dapr/proposals/pull/72/files#diff-91f410445466f13b1f41aa800ad691ac699a1a487482db8366548b14a7b72964R143). I'm not sure whether the error should be a generic error which lives outside the `oneof`s, or needs to be specific to each `oneof` API type. I'm assuming it needs to be inside each `oneof` message so that it can be fully typed.

Reply:
Especially as this channel is potentially open for a little while, it'd be better to have a `oneof` in place so the SDK can understand if the error is transient and something that just needs the channel re-opened for or if it's more of a fatal error (and should just be logged and given up on).

I'll have to find the issue, but it's a complaint on one of the recent building blocks that when there's a failure, the SDK doesn't handle it really well because there's no real information from the runtime what to do about it except that something went wrong. It'd be great to correct that scenario here.